File: python/src/palimpsest/contracts/gate_report.py
from __future__ import annotations

import argparse
import json
from typing import Any

from .gate_report_core import predeclaration_digest, predeclaration_projection
from .validation import ValidationVerdict, validate_value


def validate_gate_report(report: dict[str, Any]) -> ValidationVerdict:
    return validate_value("gate-report", report)


def complete_gate_report(
    predeclared: dict[str, Any],
    completion: dict[str, Any],
) -> dict[str, Any]:
    verdict = validate_gate_report(predeclared)
    if not verdict.accepted or predeclared.get("state") != "predeclared":
        raise ValueError("Only a valid predeclared gate report can be completed.")
    return {
        **completion,
        **predeclaration_projection(predeclared),
        "state": "completed",
        "predeclarationDigest": predeclaration_digest(predeclared),
    }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--predeclaration-digest")
    args = parser.parse_args()
    if args.predeclaration_digest is None:
        parser.error("--predeclaration-digest is required")
    print(predeclaration_digest(json.loads(args.predeclaration_digest)))


if __name__ == "__main__":
    main()

File: packages/contracts/src/gate-report.ts
import { predeclarationDigest, predeclarationProjection } from "./gate-report-core.js";
import { validateValue, type ValidationVerdict } from "./validation.js";

export interface GateCompletion {
  analysis: unknown;
  environment: unknown;
  followUp: unknown;
  producerVersions: unknown;
  rawArtifacts: unknown;
  result: unknown;
}

export function validateGateReport(report: Record<string, unknown>): ValidationVerdict {
  return validateValue("gate-report", report);
}

export function completeGateReport(
  predeclared: Record<string, unknown>,
  completion: GateCompletion,
): Record<string, unknown> {
  const verdict = validateGateReport(predeclared);
  if (!verdict.accepted || predeclared.state !== "predeclared") {
    throw new Error("Only a valid predeclared gate report can be completed.");
  }
  return {
    ...completion,
    ...predeclarationProjection(predeclared),
    state: "completed",
    predeclarationDigest: predeclarationDigest(predeclared),
  };
}

export { predeclarationDigest, predeclarationProjection };
